drivers/sensors/as5048a.c

```diff
@@ -0,0 +1,632 @@
+
+#include <nuttx/config.h>
+
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+#include <stdlib.h>
+
+#include <nuttx/kmalloc.h>
+#include <nuttx/fs/fs.h>
+#include <nuttx/spi/spi.h>
+#include <nuttx/sensors/as5048a.h>
+
+#if defined(CONFIG_SENSORS_AS5048A)
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct as5048a_dev_s
+{
+  struct qe_lowerhalf_s lower; /* AS5048A quadrature encoder lower half */
+  FAR struct spi_dev_s  *spi;  /* SPI interface */
+
+  /* Since multiple AS5048A can be connected to the same SPI bus we need
+   * to use multiple spi device ids which are employed by NuttX to select/
+   * deselect the desired AS5048A chip via their chip select inputs.
+   */
+
+  int spi_devid;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int as5048a_writeu16(FAR struct as5048a_dev_s *priv,
+                            uint16_t regaddr, uint16_t regval);
+static int as5048a_exchange(FAR struct as5048a_dev_s *priv,
+                            uint16_t regaddr, FAR uint16_t *regval);
+static uint16_t calc_even_parity(uint16_t value);
+static int as5048a_readzero(FAR struct as5048a_dev_s *priv,
+                            FAR uint16_t *zero);
+static int as5048a_writezero(FAR struct as5048a_dev_s *priv, uint16_t zero);
+static int as5048a_readagc(FAR struct as5048a_dev_s *priv,
+                           FAR uint16_t *agc);
+static int as5048a_readdiag(FAR struct as5048a_dev_s *priv,
+                            FAR uint16_t *diag);
+static int as5048a_readmag(FAR struct as5048a_dev_s *priv,
+                           FAR uint16_t *mag);
+static int as5048a_readang(FAR struct as5048a_dev_s *priv,
+                           FAR uint16_t *ang);
+
+/* Character Driver Methods */
+
+static int as5048a_setup(FAR struct qe_lowerhalf_s *lower);
+static int as5048a_shutdown(FAR struct qe_lowerhalf_s *lower);
+static int as5048a_position(FAR struct qe_lowerhalf_s *lower,
+                            FAR int32_t *pos);
+static int as5048a_reset(FAR struct qe_lowerhalf_s *lower);
+static int as5048a_ioctl(FAR struct qe_lowerhalf_s *lower, int cmd,
+                         unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct qe_ops_s g_qeops =
+{
+  .setup     = as5048a_setup,
+  .shutdown  = as5048a_shutdown,
+  .position  = as5048a_position,
+  .setposmax = NULL,            /* not supported yet */
+  .reset     = as5048a_reset,
+  .setindex  = NULL,            /* not supported yet */
+  .ioctl     = as5048a_ioctl
```

Comment:
C89 incompatible

Reply:
I'm have changed it to C89, but i don't quite understand this nuttx standard, because there are a lot of such C99 standard usages in nuttx, for example: the initialization of g_uart_ops in all arch e.g :
https://github.com/apache/incubator-nuttx/tree/master/arch/arm/src/stm32/stm32_serial.c
https://github.com/apache/incubator-nuttx/tree/master/arch/mips/src/pic32mz/pic32mz_serial.c
https://github.com/apache/incubator-nuttx/tree/master/arch/risc-v/src/esp32c3/esp32c3_serial.c

Reply:
The code in common folder has C89 requirement. The platform code can use features supported by platform compiler

Reply:
Got it, thanks for your clarification